[riccoljy] iP

riccoljy opened the pull request:
# Sentinel Chatbot PR

This PR introduces several new features and improvements to the **Sentinel** chatbot project inspired by the Duke boilerplate. 🛠️

## Key Changes:
- Added **task management** to handle _todos_, _deadlines_, and _events_.
- Enhanced command-line output with ASCII art, where you will be greeted with
```
Hello from
 _____                                                                                      _____ 
( ___ )                                                                                    ( ___ )
 |   |~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~|   | 
 |   |                                                                                      |   | 
 |   |   ________  _______   ________   _________  ___  ________   _______   ___            |   | 
 |   |  |\   ____\|\  ___ \ |\   ___  \|\___   ___\\  \|\   ___  \|\  ___ \ |\  \           |   | 
 |   |  \ \  \___|\ \   __/|\ \  \\ \  \|___ \  \_\ \  \ \  \\ \  \ \   __/|\ \  \          |   | 
 |   |   \ \_____  \ \  \_|/_\ \  \\ \  \   \ \  \ \ \  \ \  \\ \  \ \  \_|/_\ \  \         |   | 
 |   |    \|____|\  \ \  \_|\ \ \  \\ \  \   \ \  \ \ \  \ \  \\ \  \ \  \_|\ \ \  \____    |   | 
 |   |      ____\_\  \ \_______\ \__\  \__\   \ \__\ \ \__\ \__\\ \__\ \_______\ \_______\  |   | 
 |   |     |\_________\|_______|\|__| \|__|    \|__|  \|__|\|__| \|__|\|_______\|_______|   |   | 
 |   |     \|_________|                                                                     |   | 
 |   |                                                                                      |   | 
 |___|~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~|___| 
(_____)                                                                                    (_____)
```

## New Features:
1. **Task Management**:
    - Allows users to add, list, and mark tasks as done.
2. **Event Management**:
    - Schedule events with specified start and end times.
3. **Deadline Tracking**:
    - Set deadlines for tasks with a specific date.
      
## Setup Instructions:
1. Ensure your system is using **JDK 17**.
2. Configure your IDE to use the project-specific JDK:
   - Navigate to `Project Structure` → `SDKs`.
   - Set **Project language level** to _SDK default_.

Feel free to review the code and let me know if you have any suggestions. Check out the full setup guide [here](https://www.jetbrains.com/help/idea/sdk.html#set-up-jdk).

>**Notes:**
>- The main class to run the application is `Sentinel.java`, where the following `main` function is run
```Sentinel.java
public static void main(String[] args) {
        new Sentinel().run();
    }
```
>- Some commands like `todo <task>` and `deadline <task> /by <date>` are handled via inline command parsing.
>- Commands like `deadline` and `event` require date parsing either in a specific format or Internet access

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

riccoljy committed:
Add AI.md, rename chatbot, Implemented greet and exit

Renamed from Duke to Sentinel

riccoljy committed:
Implement Echo

Program repeats back whatever input

riccoljy committed:
Implement add, list

riccoljy committed:
+ Mark Done/Undone & Task.class

Implemented Task class that is used by main Sentinel file with methods like toString, setDone, setUndone, etc.

riccoljy committed:
Implement ToDo, Event, Deadline

+ 3 subclasses of Task
+ Formatting of toString/listedString
+ Slight refactor of input

riccoljy committed:
Add semi-automated testing

Used runtest.bat and adjusted input.txt and EXPECTED.TXT for testing I/O.

riccoljy committed:
Update README.md and refactor Sentinel.java

Adjusted README for Sentinel and allowed for upper/lowercase commands

riccoljy committed:
Update README.md formatting

Corrected wrong scss text

riccoljy committed:
Add /help and error-handling

Accounted for errors; KIV SentinelExceptions.java.

riccoljy committed:
Change implementation using ArrayList

riccoljy committed:
Refactor and implement delete

Combined cases mark, unmark, delete, and implemented delete using ArrayList.remove

Slight change to Task.listedString

riccoljy committed:
Use Enums

Implement Enumerations (enum Command)
Amended grammatical mistakes with plurality

riccoljy committed:
Adjust Automated Test

Added more test cases and adjusted whitespacing

riccoljy committed:
Implement Save

Create helper Classes FileLoader, FileWriter, TaskToJSONHelper for saving to .properties file.
Change Task to abstract class.

riccoljy committed:
Refactor switch case FileLoader.java

riccoljy committed:
Remove TaskToJSONHelper.java

riccoljy committed:
Merge branch 'branch-Level-7'

riccoljy committed:
Start AI implementation

riccoljy committed:
Refactor AI Implementation

Complete Query & Formatting Date/Time for deadline/events

riccoljy committed:
Implement LocalDateTime & Refactor AI Implementation

Complete Query & Formatting Date/Time for deadline/events

riccoljy committed:
Debug save Marked/Unmarked

riccoljy committed:
Update .gitignore to include .env

riccoljy committed:
Start OOP-ifying

Create Ui class

riccoljy committed:
KIV OOP & Restore missing methods from previous commit

riccoljy committed:
Major refactor OOP

Create classes Parser & Command and its subclasses.
Major refactor Sentinel and Ui to be more concise and use implemented classes.

riccoljy committed:
Minor refactor

Removed unused imports, refactor switch case blocks.

riccoljy committed:
Fix blank taskname bug

riccoljy committed:
Refactor directory and package

Package into separate packages

riccoljy committed:
Merge branch 'A-Packages'

riccoljy committed:
Merge branch 'master' into add-gradle-support

riccoljy committed:
Run Sentinel with Gradle

riccoljy committed:
Merge branch 'add-gradle-support'

riccoljy committed:
Write JUnit Test

riccoljy committed:
Refactor & Write Header Comments

riccoljy committed:
Update load/save directory, compile to jar

riccoljy committed:
Update README

riccoljy committed:
Refactor

Change variable names and whitespacing to adhere to coding standards

riccoljy committed:
Implement Find command

riccoljy committed:
Merge branch 'branch-A-CodingStandard' of https://github.com/riccoljy/ip

riccoljy committed:
Merge branch 'master' of https://github.com/riccoljy/ip into branch-Level-9

riccoljy committed:
Merge branch 'master' of https://github.com/riccoljy/ip into branch-A-JavaDoc

riccoljy committed:
Merge branch 'branch-A-JavaDoc' of https://github.com/riccoljy/ip

riccoljy committed:
Merge branch 'master' of https://github.com/riccoljy/ip into branch-Level-9

riccoljy committed:
Merge branch 'branch-Level-9' of https://github.com/riccoljy/ip

riccoljy committed:
Merge branch 'master' of https://github.com/riccoljy/ip into add-gradle-support

riccoljy committed:
Refactor minor text bugs

riccoljy committed:
WIP Gradle checkstyle

riccoljy committed:
Merge branch 'master' of https://github.com/riccoljy/ip into add-gradle-support

ITLimJiaWei commented on src/main/java/sentinel/command/FindCommand.java:
```diff
@@ -0,0 +1,17 @@
+    @Override
+    public void execute(String input) {
+        String keyword = Parser.parseKeyword(input);
+        ui.showFileredList(list, keyword);
```
Slight typo for "showFileredList", should be showFilteredList 👍

riccoljy replied:
Must have missed it doing it blurry-eyed in the middle of the night, haha! Thanks!

ITLimJiaWei commented on src/main/java/sentinel/ui/Ui.java:
```diff
@@ -172,6 +172,20 @@ public void showList(SentinelList lst) {
+     *
+     * @param lst The list of tasks to filter from.
+     */
+    public void showFileredList(SentinelList lst, String keyword) {
```
I like how you created a method dedicated to showing filtered list instead of simply reusing the usual show list method.

ITLimJiaWei commented on src/test/java/sentinel/utils/ParserTest.java:
```diff
@@ -0,0 +1,72 @@
+
+    @BeforeEach
+    public void setUp() {
+        uiMock = mock(Ui.class);
```
Could consider mocking the Parser class as well

ITLimJiaWei commented on src/main/java/sentinel/command/Command.java:
```diff
@@ -44,6 +44,7 @@ public Command(Ui ui, SentinelList list) {
     public static Command createCommand(Sentinel.CommandType commandType, Ui ui, SentinelList list) {
         return switch (commandType) {
             case list -> new ListCommand(ui, list);
+            case find -> new FindCommand(ui, list);
```
I think its your coding of switch cases is good but the coding standards specify the following format instead 🤔

``` java
switch (condition) {
case ABC:
    statements;
    // Fallthrough
case DEF:
    statements;
    break;
case XYZ:
    statements;
    break;
default:
    statements;
    break;
}



```

riccoljy replied:
Thank you for highlighting this! I'd changed it on a more recent commit after someone brought it up on the forums but haven't opened a PR yet. Thank you for reminding me!

ITLimJiaWei commented on src/test/java/sentinel/utils/SentinelListTest.java:
```diff
@@ -32,13 +32,13 @@ public void testSizeOneWhenSizeIsOne() {
     public void testSizeOneWhenSizeIsNotOne() {
         // Act & Assert
-        assertFalse(sentinelList.sizeOne());
+        assertFalse(sentinelList.isSizeOne());
```
Good work adhering to the coding standard here and other areas where you made this change 👍

ITLimJiaWei commented on src/test/java/sentinel/task/ToDoTest.java:
```diff
@@ -0,0 +1,67 @@
+
+    @Test
+    public void testConstructor() {
+        String description = "Test ToDo";
```
You could consider having ``ToDo todo = new ToDo("Test ToDo");` declared at the class level so you won't have to repeat it for every JUnit test that utilizes it.

riccoljy replied:
Oh thank you for your suggestion. I was actually hesitant to do that due to the modifications I might make in each new test (eg: First test I just want to check on markDone and second test I want to test markUndone (for newly initialised ToDo instance that is already marked undone).

ITLimJiaWei reviewed (commented):
LGTM. Just minor nits to note/fix.

riccoljy committed:
WIP Gradle checkstyle

riccoljy committed:
Merge branch 'add-gradle-support' of https://github.com/riccoljy/ip

curtischang2510 commented on src/main/java/sentinel/api/GeminiApi.java:
```diff
@@ -0,0 +1,101 @@
+     * @return The text result from Gemini.
+     */
+
+    public static String query(String input) {
```
Very interesting approach. Hopefully, it does not take too long to query the API if not the app might seem unresponsive :). 

Also have interested to find out if there is any fall back if user is not connected to the internet?

riccoljy replied:
Currently in my latest commit, users without internet or `RequestTimeOut`s will fallback on the standard `LocalDateTime.parse(<String input>)` method where users have to abide by the format `YYYY-MM-DDTHH:M:S`. Thanks for asking!

curtischang2510 commented on src/main/java/sentinel/command/Command.java:
```diff
@@ -0,0 +1,55 @@
+ */
+public abstract class Command {
+    protected Ui ui;
+    protected SentinelList list;
```
Name could be more descriptive. Maybe something like `taskList` or `userTasks`  :)

curtischang2510 commented on src/main/java/sentinel/storage/FileWriter.java:
```diff
@@ -0,0 +1,82 @@
+    private static final Path DIRECTORY_PATH = Paths.get("sentinel-saves");
+    private static final Path FILE_PATH = DIRECTORY_PATH.resolve("tasks.properties");
+
+    private final SentinelList arrayList;
```
arrayList is probably not the best name choice here 😅 . As above, `userTasks` or a more descriptive name could be helpful :)

curtischang2510 commented on src/main/java/sentinel/ui/Ui.java:
```diff
@@ -0,0 +1,226 @@
+                   |___|~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~|___|\s
+                  (_____)                                                                                    (_____)""";
+
+    private static final String helpText = """
```
That's really nice

curtischang2510 commented on src/main/java/sentinel/utils/Parser.java:
```diff
@@ -0,0 +1,154 @@
+            return null;
+        }
+
+        switch (commandType) {
```
Case indentation does not follow course guidelines here 😢

riccoljy replied:
Thank you for highlighting this! I'd changed it on a more recent commit after someone brought it up on the forums but haven't opened a PR yet. Thank you for reminding me!

curtischang2510 reviewed (commented):
Overall code looks really clean. Great job man

riccoljy committed:
Refactor to follow Java Style Guide

riccoljy committed:
Implement GUI for Sentinel

riccoljy committed:
Merge branch 'GUI' of https://github.com/riccoljy/ip

riccoljy committed:
Utilise Varags

riccoljy committed:
Add assert statements to document key assumptions

This commit introduces the use of the `assert` feature to document important assumptions at various points in the code. These assertions serve as sanity checks to ensure that critical conditions hold true during execution, helping to catch unexpected states early.

- Added assertions to key methods to verify preconditions, postconditions, and invariants.
- Assertions are used to catch logical errors during development and improve code reliability.
- Assertions are disabled in production for performance, as they are intended for debugging and testing phases.

This approach is distinct from JUnit assertions, which are used for unit testing.

riccoljy committed:
refactor: Improve code quality by reducing nesting and clarifying happy path

This commit refactors the code to enhance its overall quality by addressing key areas that promote better readability and maintainability. The changes are guided by established 'Code Quality' principles rather than personal intuition.

- Reduced deep nesting in several areas of the code, minimizing "arrowhead" structures.
- Restructured code to make the happy path—the normal execution path—clearer and more prominent.
- Nested 'unusual' or error cases to simplify the main logic flow.
- Ensured that every switch/case statement includes a default branch to handle unexpected scenarios.

These improvements aim to make the code easier to understand, debug, and maintain in the long term.

riccoljy committed:
refactor: Replace for loops with Java Streams for cleaner and more functional code

This commit refactors the code to leverage Java's Streams API, improving readability and conciseness by replacing most traditional for loops with stream operations.

- Replaced standard for loops with streams to handle iterations more declaratively.
- Used stream operations such as `filter`, `map`, and `collect` to transform collections and process data.
- Simplified complex loop logic, making the code more functional and easier to reason about.
- Ensured no change in functionality or performance regressions after the refactor.

The use of streams enhances code clarity and aligns with modern Java best practices.

riccoljy committed:
Merge branch 'branch-A-Assertions' of https://github.com/riccoljy/ip

riccoljy committed:
Merge branch 'master' of https://github.com/riccoljy/ip into branch-A-CodeQuality

riccoljy committed:
Merge branch 'branch-A-CodeQuality' of https://github.com/riccoljy/ip

riccoljy committed:
Merge branch 'master' of https://github.com/riccoljy/ip into branch-A-Streams

# Conflicts:
#	src/main/java/sentinel/utils/SentinelList.java

riccoljy committed:
Merge branch 'branch-A-Streams' of https://github.com/riccoljy/ip

riccoljy committed:
Add gradle.yml workflow to Github Actions

riccoljy committed:
Update chatbot icons

riccoljy committed:
Update chatbot icons and error message UI

riccoljy committed:
Update for Checkstyle

riccoljy committed:
Change Sentinel responses

riccoljy committed:
Merge pull request #2 from riccoljy/branch-A-BetterGui

Branch a better gui

riccoljy committed:
Merge branch 'branch-A-Personality' into master

riccoljy committed:
Merge pull request #3 from riccoljy/master

Pull master into branc-A-Personality

riccoljy committed:
Merge pull request #4 from riccoljy/branch-A-Personality

Merge pull request #3 from riccoljy/master

riccoljy committed:
Merge pull request #5 from riccoljy/master

Pull master into branch-A-MoreTesting

riccoljy committed:
Update JUnit Test for Sentinel responses

riccoljy committed:
Merge pull request #6 from riccoljy/branch-A-MoreTesting

Branch a more testing

riccoljy committed:
Debug AI Parsing & Sentinel Responses

riccoljy committed:
Add Ui.png and update README.md

riccoljy committed:
Remove ununsed import

riccoljy committed:
Merge pull request #7 from riccoljy/branch-A-UserGuide

Branch a user guide

riccoljy committed:
Debug misdirected file

riccoljy committed:
WIP GeminiException

riccoljy committed:
Rely less Gemini API

riccoljy committed:
Change Gemini error message to specify date format
